mkosi/machine.py

```diff
@@ -122,9 +142,71 @@ def run(self, commands: Sequence[str], timeout: int = 900, check: bool = True) -
 
         return process
 
-    def __exit__(self, *args: Any, **kwargs: Any) -> None:
+    def _kill(self, *args: Any, **kwargs: Any) -> None:
         if self._serial:
             self._serial.kill(signal.SIGTERM)
             self.exit_code = self._serial.wait()
             self._serial = None
         self.stack.__exit__(*args, **kwargs)
+
+    def __exit__(self, *args: Any, **kwargs: Any) -> None:
+        self._kill(*args, **kwargs)
+
+    def reboot(self) -> None:
+        self._kill(None, None, None)
+        self._boot()
+
+
+class MkosiMachineTest(unittest.TestCase):
+    mkosi_tests_no_qemu: bool
+    mkosi_tests_no_nspawn: bool
+    args: Sequence[str]
+    machine: Machine
+    exc_info: Tuple[Optional[Type[BaseException]], Optional[BaseException], Optional[TracebackType]] \
+              = (None, None, None)
+
+    def _callTestMethod(self, method: str) -> None:
+        try:
+            super()._callTestMethod(method) # type: ignore
+        except Exception as e:
+            self.exc_info = exc_info()
+            raise e
+
+    # Mirroring the logic from https://github.com/systemd/systemd/blob/main/src/basic/parse-util.c.
+    @classmethod
+    def _parse_boolean(cls, environment_var: str) -> bool:
+        value = os.getenv(environment_var)
+        if not value:
+            die(f"No value assigned to {environment_var}.")
+
+        if value in ("1", "yes", "y", "true", "t", "on"):
+            return True
+        if value in ("0", "no", "n", "false", "f", "off"):
+            return False
+
+        die(f"No valid value assigned to {environment_var}.")
+
+    def __init_subclass__(cls, args: Sequence[str] = []) -> None:
+        cls.args = args
+
+    @classmethod
+    def setUpClass(cls) -> None:
+        cls.machine = Machine(cls.args)
+
+        machine_verb = cls.machine.args.verb
+        no_nspawn = cls._parse_boolean("MKOSI_TESTS_NO_NSPAWN")
+        no_qemu = cls._parse_boolean("MKOSI_TESTS_NO_QEMU")
```

Comment:
The way this is setup, both environment variables need values in order to execute tests.
Maybe if we have no assigned value we should just raise a warning message? As the number of tests pile up it would be important to make sure such variables are defined though.